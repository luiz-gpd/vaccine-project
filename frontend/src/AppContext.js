import { createContext } from 'react'

const AppContext = createContext();


// const reducer = (state, action) => {
//     switch (action.type) {
//         case "EXEMPLE": {
//             return {
//                 ...state,
//                 loggedUser: action.payload,
//             };
//         }

//         default: {
//             return state;
//         }
//     }
// }

export default AppContext