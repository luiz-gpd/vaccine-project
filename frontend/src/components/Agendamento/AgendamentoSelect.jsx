import React from 'react'
import { ClaySelect } from '@clayui/form'

const AgendamentoSelect = () => {

    const options = [
        {   label: "8:00", value: "1" },
        {   label: "8:30", value: "2" },
        {   label: "9:00", value: "3" },
        {   label: "9:30", value: "4" },
        {   label: "10:00", value: "5" },
        {   label: "10:30", value: "6" },
        {   label: "11:00", value: "7" },
        {   label: "11:30", value: "8" },
        {   label: "12:00", value: "9" },
        {   label: "12:30", value: "10" },
        {   label: "13:00", value: "11" },
        {   label: "13:30", value: "12" },
        {   label: "14:00", value: "13" },
        {   label: "14:30", value: "14" },
        {   label: "15:00", value: "15" },
        {   label: "15:30", value: "16" },
        {   label: "16:00", value: "17" },
        {   label: "16:30", value: "18" },
        {   label: "17:00", value: "19" },
        {   label: "17:30", value: "20" },
    ];

    return (
        <ClaySelect aria-label="Select Label" id="mySelectId">

            {options.map(option => (
                <ClaySelect.Option
                    key={option.value}
                    label={option.label}
                    value={option.value}
                />
            ))}

        </ClaySelect>
    )
}
export default AgendamentoSelect