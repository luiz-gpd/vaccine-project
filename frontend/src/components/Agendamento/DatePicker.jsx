import React from 'react'
import ClayDatePicker from '@clayui/date-picker'
import { Field, ErrorMessage } from 'formik'

const DatePicker = (props) => {
    const { b, name, ...otherProps } = props
    return (
        <div>
            <b>{b}</b>
            <Field name={name}>
                {({ form, field }) => {
                    const { setFieldValue } = form
                    const { value } = field
                    return (
                        <ClayDatePicker
                            name={name}
                            dateFormat='dd/MM/yyyy'
                            value={value}
                            onValueChange={val => setFieldValue(name, val)}
                            weekdaysShort={['D', 'S', 'T', 'Q', 'Q', 'S', 'S']}
                            months={['Janeiro', 'Feveiro', 'MarÃ§o',
                                'Abril', 'Maio', 'Junho', 'Julho', 'Agosto',
                                'Setembro', 'Outubro', 'Novembro', 'Dezembro']}
                            {...otherProps}
                        />
                    )
                }}
            </Field>
            <ErrorMessage name={name} />
        </div>
    )
}
export default DatePicker
