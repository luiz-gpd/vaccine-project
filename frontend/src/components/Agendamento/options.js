const options = [
    { label: "Selecione um hor√°rio", value: "0" },
    { label: "8:00", value: "8:00" },
    { label: "9:00", value: "9:00" },
    { label: "10:00", value: "10:00" },
    { label: "11:00", value: "11:00" },
    { label: "12:00", value: "12:00" },
    { label: "13:00", value: "13:00" },
    { label: "14:00", value: "14:00" },
    { label: "15:00", value: "15:00" },
    { label: "16:00", value: "16:00" },
    { label: "17:00", value: "17:00" },
];


export default options;