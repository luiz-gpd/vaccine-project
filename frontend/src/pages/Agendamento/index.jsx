import React, { useContext, useState, useEffect } from 'react'
import { Prompt } from 'react-router'
import Page from '../../components/Page'
import { Formik, ErrorMessage, Form } from 'formik'
import schema from '../../utils/schema'
import ClayForm, { ClayInput, ClaySelectWithOption } from '@clayui/form'
import ClayButton from '@clayui/button'
import options from '../../components/Agendamento/options'
import AppContext from '../../AppContext'
import DatePicker from '../../components/Datepicker'
import api from '../../utils/api'
import Toast from '../../components/Toast'


const Agendamento = () => {
    const [{ toast }, dispatch] = useContext(AppContext);
    const [toastType, setToastType] = useState(true);

    useEffect(() => {
        window.onbeforeunload = function () {
          return true;
        };
        return () => {
          window.onbeforeunload = null;
        };
      }, []);

    const onSubmit = async (values) => {

        const bornOn = values.birthDate;
        const d = new Date();
        var idade = d.getFullYear() - bornOn.getYear() - 1900;
        if (d.getMonth() + 1 < bornOn.getMonth() + 1 || (d.getMonth() + 1 === bornOn.getMonth() + 1 && d.getDate() < bornOn.getDate())) {
            idade--;
        }

        const user = {
            name: values.name,
            age: idade,
            consultationDate: values.consultationDate,
            consultationTime: values.consultationTime,
            attended: false,
            consultInfo: ""
        };

        try {
            const response = await api.post('/user', user)
            const value = Math.random() * 100
            if (response.data === "Já há duas pessoas nesse horário") {
                setToastType(false);
                dispatch({
                    type: "SHOW_TOAST",
                    payload: { value },
                  });
                } else {
                    setToastType(true);
                    dispatch({
                        type: "SHOW_TOAST",
                        payload: { value },
                      }); 
            }
        } catch (e) {
            alert(e.response.data.message);
        }
    }

    return (
        <Page title="Realize seu Agendamento">
            <Formik
                validationSchema={schema}
                onSubmit={onSubmit}
                initialValues={{
                    name: '',
                    birthDate: '',
                    consultationDate: '',
                    consultationTime: '',
                }}>
                {({ values, handleChange, handleSubmit, dirty }) => (
                    <Form className="mt-6" onSubmit={handleSubmit}>
                        <Prompt 
                        when={dirty}
                        message="Você tem alterações não salvas, deseja mesmo mesmo sair?"/>
                        <ClayForm.Group>
                            <b>Nome:</b>
                            <ClayInput name="name" type="text" value={values.name} placeholder="Exemplo exemplo" onChange={handleChange} />
                            <ErrorMessage name="name" />
                        </ClayForm.Group>
                        <div className="form-group-autofit"><div className="mr-6">
                            <ClayForm.Group>
                                <DatePicker
                                    b="Data de Nascimento: "
                                    name="birthDate"
                                    maxDate={new Date()}
                                />
                            </ClayForm.Group></div>
                            <div className="ml-6">
                                <ClayForm.Group>
                                    <DatePicker
                                        b="Data da consulta: "
                                        name="consultationDate"
                                        minDate={new Date()}
                                    />
                                </ClayForm.Group></div></div>
                        <ClayForm.Group>
                            <b>Horário da Consulta:</b>
                            <ClaySelectWithOption
                                name="consultationTime"
                                onChange={handleChange}
                                options={options}
                            >
                            </ClaySelectWithOption>
                            <ErrorMessage name="consultationTime" />
                        </ClayForm.Group>
                        <ClayButton type="submit">Enviar</ClayButton>
                    </Form>
                )}
            </Formik>
            <Toast title={toastType ? "Pronto!" : "Erro"}
            onClose={() => {
                dispatch({
                    type: "HIDE_TOAST",
                    payload: { data: "" },
                });
            }}
            toast={toast}
            type={toastType ? "success" : "danger"}
            >{toastType ? "Cadastro realizado com successo!" : "Já há duas pessoas nesse horário!"}</Toast>
        </Page> 
    )
}

export default Agendamento;