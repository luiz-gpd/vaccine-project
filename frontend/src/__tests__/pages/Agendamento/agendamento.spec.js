import Agendamento from "../../../pages/Agendamento"
import AppContext, { initialState } from "../../../AppContext"
import { render, fireEvent, waitFor } from "@testing-library/react"
import { BrowserRouter } from "react-router-dom"

const AgendamentoWrapper = ({ state = initialState, dispatch }) => (
    <AppContext.Provider value={[state, dispatch]}>
        <BrowserRouter>
            <Agendamento/>
        </BrowserRouter>
    </AppContext.Provider>
);

describe("Agendamento page", () => {
    it("renders", () => {
        const { asFragment } = render(<AgendamentoWrapper />);

        expect(asFragment()).toMatchSnapshot();
    });
    
    it("Button should submit when clicked", async () => {
        const { container } = render(<AgendamentoWrapper />);

        const formButton = container.querySelector(
            'button[type="submit"]'
          );
        
        await waitFor(() => {
          fireEvent.click(formButton);
        })
    });
});