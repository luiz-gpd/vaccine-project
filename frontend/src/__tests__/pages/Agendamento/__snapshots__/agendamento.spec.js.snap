// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Agendamento page renders 1`] = `
<DocumentFragment>
  <div
    class="mt-4 container"
  >
    <div
      class="row justify-content-between"
    >
      <div
        class="col"
      >
        <h1>
          Realize seu Agendamento
        </h1>
      </div>
      <div
        class="col-3"
      >
        <img
          alt="Logo do site"
          src="VaciCuidar_logo.png"
          width="250px"
        />
      </div>
    </div>
    <form
      action="#"
      class="mt-6"
    >
      <div
        class="form-group"
      >
        <b>
          Nome:
        </b>
        <input
          class="form-control"
          name="name"
          placeholder="Exemplo exemplo"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group-autofit"
      >
        <div
          class="mr-6"
        >
          <div
            class="form-group"
          >
            <div
              class="form-control"
            >
              <b
                for="birthDate"
              >
                Data de Nascimento: 
              </b>
              <br />
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container"
                >
                  <input
                    autocomplete="off"
                    class=""
                    id="birthDate"
                    name="birthDate"
                    placeholder="00/00/0000"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <br />
            </div>
          </div>
        </div>
        <div
          class="ml-6"
        >
          <div
            class="form-group"
          >
            <div
              class="form-control"
            >
              <b
                for="consultationDate"
              >
                Data da consulta: 
              </b>
              <br />
              <div
                class="react-datepicker-wrapper"
              >
                <div
                  class="react-datepicker__input-container"
                >
                  <input
                    autocomplete="off"
                    class=""
                    id="consultationDate"
                    name="consultationDate"
                    placeholder="00/00/0000"
                    type="text"
                    value=""
                  />
                </div>
              </div>
              <br />
            </div>
          </div>
        </div>
      </div>
      <div
        class="form-group"
      >
        <b>
          Horário da Consulta:
        </b>
        <select
          class="form-control"
          name="consultationTime"
        >
          <option
            value="0"
          >
            Selecione um horário
          </option>
          <option
            value="08:00"
          >
            08:00
          </option>
          <option
            value="08:30"
          >
            08:30
          </option>
          <option
            value="09:00"
          >
            09:00
          </option>
          <option
            value="09:30"
          >
            09:30
          </option>
          <option
            value="10:00"
          >
            10:00
          </option>
          <option
            value="10:30"
          >
            10:30
          </option>
          <option
            value="11:00"
          >
            11:00
          </option>
          <option
            value="11:30"
          >
            11:30
          </option>
          <option
            value="12:00"
          >
            12:00
          </option>
          <option
            value="12:30"
          >
            12:30
          </option>
          <option
            value="13:00"
          >
            13:00
          </option>
          <option
            value="13:30"
          >
            13:30
          </option>
          <option
            value="14:00"
          >
            14:00
          </option>
          <option
            value="14:30"
          >
            14:30
          </option>
          <option
            value="15:00"
          >
            15:00
          </option>
          <option
            value="15:30"
          >
            15:30
          </option>
          <option
            value="16:00"
          >
            16:00
          </option>
          <option
            value="16:30"
          >
            16:30
          </option>
          <option
            value="17:00"
          >
            17:00
          </option>
          <option
            value="17:30"
          >
            17:30
          </option>
        </select>
      </div>
      <button
        class="form-button btn btn-primary"
        type="submit"
      >
        Enviar
      </button>
    </form>
    <div
      class="alert-container container"
    >
      <div
        class="alert-notifications alert-notifications-fixed"
      />
    </div>
  </div>
</DocumentFragment>
`;
